src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Check.java

```diff
@@ -4294,13 +4256,12 @@ private boolean isCanonical(JCTree tree) {
     /** Check that an auxiliary class is not accessed from any other file than its own.
      */
     void checkForBadAuxiliaryClassAccess(DiagnosticPosition pos, Env<AttrContext> env, ClassSymbol c) {
-        if (lint.isEnabled(Lint.LintCategory.AUXILIARYCLASS) &&
-            (c.flags() & AUXILIARY) != 0 &&
+        if ((c.flags() & AUXILIARY) != 0 &&
```

Comment:
Here, and on a few other places, the checks were ordered so that the potentially more expensive checks were done after the `isEnabled` check. Checking the exact places, I don't think delaying the check is terrible for the cases in this PR, but if we would like to be adding new lints, we probably need something that will avoid running the lint's code completely in a (semi-)automatic way. That could be part of a more generic 22088.

(Admittedly, on some places, the `isEnabled` check was last, after the more expensive checks. This is particularly the case for the `TRY` lint, where the checks might be considerable.)

Reply:
> we probably need something that will avoid running the lint's code completely in a (semi-)automatic way. That could be part of a more generic 22088.

Agreed.

Slight comment hijack follows...

The [unnecessary suppression warning proposal](https://mail.openjdk.org/pipermail/compiler-dev/2024-November/028573.html) would create a natural answer for this, because there would be a new concept "active" along with "enabled": A category is active if it's enabled OR we are tracking unnecessary suppressions, the category is being suppressed at this point in the code, and so far that suppression is unnecessary (i.e., no warnings would have fired yet).

So linter code would want to check for "active" status (instead of "enabled" status) before doing a non-trivial calculation, which would end up looking like this:
```java
if (lint.isActive(LintCategory.FOO) &&
  complicatedCalculationA() &&
  complicatedCalculationB()) {
    lint.logIfEnabled(log, pos, LintWarnings.FooProblem);
}
```
But perhaps then we should then consider lambda-ification, which would simplify this down to:
```java
lint.check(log, pos, LintWarnings.FooProblem,
  () -> complicatedCalculationA() && complicatedCalculationB());
```
This could be your "(semi-)automatic" option.

Reply:
> Here, and on a few other places, the checks were ordered so that the potentially more expensive checks were done after the `isEnabled` check. Checking the exact places, I don't think delaying the check is terrible for the cases in this PR, but if we would like to be adding new lints, we probably need something that will avoid running the lint's code completely in a (semi-)automatic way. That could be part of a more generic 22088.
> 
> (Admittedly, on some places, the `isEnabled` check was last, after the more expensive checks. This is particularly the case for the `TRY` lint, where the checks might be considerable.)

I was aware of that. As you might notice, in some cases I preserved the original code structure - in other cases I've changed it when the preceding checks seemed O(1) -- I might have made some mistakes, but that was the spirit.

Reply:
> But perhaps then we should then consider lambda-ification, which would simplify this down to:
> 
> ```java
> lint.check(log, pos, LintWarnings.FooProblem,
>   () -> complicatedCalculationA() && complicatedCalculationB());
> ```

I thought about this as well but found it a bit too convoluted for my liking. One might already argue that `Lint::logIfEnabled` is doing two things at once (check and log). Making it an open box where you check if the lint is enabled, you check some user provided condition, and if both are true then you log" seems a rather ad-hoc API to add :-)

Reply:
Overall I just think the problem of "you checked for A, but are logging a B" is not big enough to warrant an hammer like this.

Reply:
> What would be better, IMHO, is to investigate an automate way to e.g. enhance something like CheckExamples to automatically test for suppression of lint warnings.

Indeed... and I've already got a [precursor](https://github.com/archiecobbs/jdk/blob/suppression/test/langtools/tools/javac/lint/SuppressionWarningTest.java) which could serve as a starting point for something like that. What's missing in the precursor is automatic verification that _every_ lint warning is tested, since until this PR that was impractical. But with this PR we can now enumerate all the `LintWarning`'s, so it would be easy to automatically verify we have complete coverage. Of course, we'd also have to add all the missing warnable code examples, which is not hard but somewhat tedious. Let me know if you are interested in putting something together, I'm happy to help.